File: 213. House Robber II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        
        int res1 = nums[0];
        vector<int> dp(3, 0);
        // 打劫第一个房子
        for (int i = 0; i < n - 1; i++) {
            int t = 0;
            for (int j = 0; j < 2; j++) 
                t = max(t, dp[j] + nums[i]);
            res1 = max(t, res1);
            dp.push_back(t);
            dp.erase(dp.begin());
        }
        if (n == 1)
            return res1;
        int res2 = nums[1];
        // 不打劫第一个房子
        dp.clear();
        dp.resize(3, 0);
        for (int i = 1; i < n; i++) {
            int t = 0;
            for (int j = 0; j < 2; j++) 
                t = max(t, dp[j] + nums[i]);
            res2 = max(t, res2);
            dp.push_back(t);
            dp.erase(dp.begin());
        }
        return(max(res1, res2));
    }
};


class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n==0)
            return 0;
        if (n==1)
            return nums[0];
        if (n==2)
            return max(nums[0], nums[1]);
        
        // 打劫第一个房子
        int dp[2] = {nums[0], max(nums[0], nums[1])};
        for (int i = 2; i < n-1; i++) {
            int t = dp[1];
            dp[1] = max(dp[1], nums[i] + dp[0]);
            dp[0] = t;
        }
        int r1 = dp[1];

        // 不打劫第一个房子
        dp[0] = nums[1];
        dp[1] = max(nums[1], nums[2]);
        for (int i = 3; i < n; i++) {
            int t = dp[1];
            dp[1] = max(dp[1], nums[i] + dp[0]);
            dp[0] = t;
        }
        int r2 = dp[1];

        return max(r1, r2);
    }
};

File: go/src/leetcode/213. House Robber II/213. House Robber II.go
/**
# 思路：考虑两种情况，由于是环形，拿了第一家就不能拿最后要一家，所以分成
#       两种情况求取最大值就行了，从后往前遍历可以避免f1, f2初始值设置的问题
 */
package main

import "fmt"

func rob(nums []int) int {
	n := len(nums)

	var steal = func(nums []int, l, r int) int {
		d1, d2, d3 := 0, 0, 0
		for i := r; i >= l; i-- {
			d3 = max(d2, d1 + nums[i])
			d1 = d2
			d2 = d3
		}
		return d2
	}

	return max(steal(nums, 0, n-2), steal(nums, 1, n-1))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(rob([]int{2,3,2}))
	fmt.Println(rob([]int{1,2,3,1}))
	fmt.Println(rob([]int{0}))
	fmt.Println(rob([]int{1,3,1,3,100}))
}